### PR TITLE
Elliptic curves utilities refactory

### PR DESCRIPTION
**Pull request by davxy**
- Usage the new published [arkworks-extensions](https://github.com/paritytech/arkworks-extensions) crates. 
  Hooks are internally defined to jump into the proper host functions.
- Conditional compilation of each curve (gated by feature with curve name)
- Separation in smaller host functions sets, divided by curve (fits nicely with prev point)

Runtime dev should not care that much about host functions and hooks as these are internally defined and he
can use the library as it'd use Arkworks upstream in his runtimes.

---

Please be aware that this library is mostly designed for internal usage within crates residing in the `polkadot-sdk` workspace, such as Bandersnatch. This is to avoid potential problems, as demonstrated in issues like https://github.com/paritytech/polkadot-sdk/issues/2013.

Devs of pallets outside the workspace can utilize this library or the stand alone crates available on `crates.io` and maintained in [arkworks-substrate](https://github.com/paritytech/arkworks-substrate) repository if they like.
Stand alone crates essentially offer the same functionality as the curve modules found here.

### CODE REVIEW

**Review by michalkucharczyk — APPROVED** [review]
I did not dive into the logic, implementation details.
Overall it looks good.

I did not notice any test, maybe we should have some, (at least for utils)?

File: substrate/primitives/crypto/ec-utils/src/lib.rs [hunk 1]
```diff
@@ -15,272 +15,39 @@
...
+//! The definitions make use of host functions to offload the non-native
+//! computational environment from the some of the most computationally
+//! expensive operations by internally leveraging the
+//! [arkworks-extensions](https://github.com/paritytech/arkworks-extensions)
+//! library.
+//!
+//! The exported types are organized and named in a way that mirrors the structure
+//! of the types in the original Arkworks library. This design choice aims to make
+//! it easier for users already familiar with the library to understand and utilize
+//! the exported types effectively.
```
Comment (michalkucharczyk) [inline substrate/primitives/crypto/ec-utils/src/lib.rs:30]: Do I understand correctly that this crate is mainly intended to use in runtime, and it provides some abstraction layer which allows to call the Elliptic curves functionality which is delegated to host?

Comment (davxy) [reply substrate/primitives/crypto/ec-utils/src/lib.rs:30]: Yes correct.

Allows to use stuff in arkworks ZK stuff in the runtime.
Our types are like upstream ones but jumps into host functions when required.

File: substrate/primitives/crypto/ec-utils/src/lib.rs [hunk 2]
```diff
@@ -15,272 +15,39 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//! Elliptic Curves host functions which may be used to handle some of the *Arkworks*
-//! computationally expensive operations.
+//! Elliptic curves which are mostly compatible with *Arkworks* library
+//! mostly useful in non-native contexts.
```
Comment (michalkucharczyk) [inline substrate/primitives/crypto/ec-utils/src/lib.rs:19]: maybe?
```suggestion
//! This crate offers elliptic curves functionality which is mostly compatible with *Arkworks* library.
//! It is primarily designed for use in non-native contexts.
```

also wouid be nice to have link to Arkworks.

File: substrate/primitives/crypto/ec-utils/src/lib.rs [hunk 3]
```diff
@@ -15,272 +15,39 @@
...
+//!
+//! The definitions make use of host functions to offload the non-native
+//! computational environment from the some of the most computationally
+//! expensive operations by internally leveraging the
+//! [arkworks-extensions](https://github.com/paritytech/arkworks-extensions)
+//! library.
```
Comment (michalkucharczyk) [inline substrate/primitives/crypto/ec-utils/src/lib.rs:25]:
```suggestion
//! The implementation make use of host functions to offload the non-native
//! environment from some of the most computationally
//! expensive operations. It internally leverages the
//! [arkworks-extensions](https://github.com/paritytech/arkworks-extensions)
//! library.
```

**Comment by davxy** [comment]
> So these crates here are just a verbatim copy?

@bkchr Not really verbatim (but in practice, at the moment, yes). Here we leverage some tiny `encode/decode` shared functions.

After this is merged [arkworks-substrate](https://github.com/paritytech/arkworks-substrate) will be refactored to just use the hooks and the stuff we are implementing here. So the crates there will act as some kind of convenience facade.

The arkworks-substrate crates are handy if someone wants to include the single crates individually and thus more closely mimic what you'd do when using upstream Arkworks.

**Review by skunert — APPROVED** [review]

**Pull request merged**